### PR TITLE
Use a hardlink copy to support AUFS (and OverlayFS)

### PR DESCRIPTION
**Pull request by pcarranzav**
Depends on resin-io-modules/docker-toolbelt#4 being merged and published as 1.3.0.

### CODE REVIEW

File: lib/docker-delta.coffee [hunk 1]
```diff
@@ -104,69 +104,86 @@ parseDeltaStream = (input) ->
 
 		input.on('readable', parser)
 
+nullDisposer = ->
+	Promise.try ->
```
Comment (Page-) [inline lib/docker-delta.coffee]: I think you can just use `Promise.resolve(null)`

File: lib/docker-delta.coffee [hunk 2]
```diff
@@ -104,69 +104,86 @@ parseDeltaStream = (input) ->
...
-			switch dockerDriver
-				when 'btrfs'
-					if srcRoot?
-						btrfs.deleteSubvolAsync(dstRoot)
-						.then ->
-							btrfs.snapshotSubvolAsync(srcRoot, dstRoot)
-				when 'overlay'
-					if srcRoot?
-						rsyncArgs.push('--link-dest', srcRoot)
+		Promise.all [
```
Comment (Page-) [inline lib/docker-delta.coffee]: Can you turn this into a `Promise.join`

File: lib/docker-delta.coffee [hunk 3]
```diff
@@ -104,69 +104,86 @@ parseDeltaStream = (input) ->
...
+			docker.infoAsync().get('Driver')
+			dstIdPromise
+			dstIdPromise.then(docker.imageRootDir)
+		]
+		.spread (dockerDriver, dstId, dstRoot) ->
+			# trailing slashes are significant for rsync
+			dstRoot = path.join(dstRoot, '/')
+
+			Promise.attempt ->
```
Comment (Page-) [inline lib/docker-delta.coffee]: Can you use `Promise.try` instead of the `attempt` alias please

File: lib/docker-delta.coffee [hunk 4]
```diff
@@ -104,69 +104,86 @@ parseDeltaStream = (input) ->
...
+				# rsync doesn't fsync by itself
+				utils.waitPidAsync(spawn('sync'))
+			.then ->
+				deltaStream.emit('id', dstId)
+		.catch (e) ->
+			# If the process failed for whatever reason, cleanup the empty image
+			dstIdPromise.then (dstId) ->
+				docker.getImage(dstId).removeAsync()
+				.catch (e) ->
+					deltaStream.emit('error', e)
```
Comment (Page-) [inline lib/docker-delta.coffee:184]: The cleanup error will get emitted before the root error, so the root error might well get ignored - it's probably worth emitting in the order they happen?

**Review by Page- — CHANGES_REQUESTED** [review]

File: lib/docker-delta.coffee (hunk 4 above)
Comment (pcarranzav) [reply lib/docker-delta.coffee:184]: That's a good point, this is the original way it was done but I'll change it so that it's in the correct order.

Comment (pcarranzav) [reply lib/docker-delta.coffee:184]: Done.

File: lib/docker-delta.coffee (hunk 1 above)
Comment (pcarranzav) [reply lib/docker-delta.coffee]: Done.

File: lib/docker-delta.coffee (hunk 2 above)
Comment (pcarranzav) [reply lib/docker-delta.coffee]: Done.

File: lib/docker-delta.coffee (hunk 3 above)
Comment (pcarranzav) [reply lib/docker-delta.coffee]: Done.

**Review by Page- — APPROVED** [review]

**Pull request merged**